Error Handling: Deploying to Connect while it's unavailable throws ambiguous error

Deploying to a Connect Instance that is not available throws an ambiguous error:

```> bin/darwin/arm64/1.0.alpha2-189-gba921678/publisher deploy test/content/bundles/stock-dashboard-python/
Deploy to server http://localhost:3939 using account env and configuration default, creating deployment Untitled-1
Check Configuration...             time=2024-02-27T17:04:59.854-05:00 level=WARN msg="failed to write updated deployment record" name="" err="toml: cannot encode value of type uintptr"
[ERROR] unknown (Source=%!q(<nil>), Addr="[::1]:3939", Err="connect: connection refused", Op="dial", Net="tcp", dashboardUrl="", url="")

dial tcp [::1]:3939: connect: connection refused.
```

Specifically: ```toml: cannot encode value of type uintptr```

This was hit when using an unavailable Docker image as my env account.

--- a/internal/clients/http_client/http_client.go
+++ b/internal/clients/http_client/http_client.go
@@ -82,7 +82,7 @@ func (c *defaultHTTPClient) do(method string, path string, body io.Reader, bodyT
 		if e, ok := err.(net.Error); ok && e.Timeout() {
 			return nil, types.NewAgentError(events.OperationTimedOutCode, err, nil)
 		}
-		return nil, err
+		return nil, types.NewAgentError(events.ConnectionFailedCode, err, nil)
 	}
 	defer resp.Body.Close()
 

--- a/internal/events/error_codes.go
+++ b/internal/events/error_codes.go
@@ -11,6 +11,7 @@ const (
 	AuthenticationFailedCode  ErrorCode = "authFailedErr"            // Couldn't authenticate to publishing server
 	PermissionsCode           ErrorCode = "permissionErr"            // Server responded with 403 forbidden
 	OperationTimedOutCode     ErrorCode = "timeoutErr"               // HTTP request to publishing server timed out
+	ConnectionFailedCode      ErrorCode = "connectionFailed"         // Couldn't connect to Connect
 	ServerErrorCode           ErrorCode = "serverErr"                // HTTP 5xx code from publishing server
 	VanityURLNotAvailableCode ErrorCode = "vanityURLNotAvailableErr" // Vanity URL already in use
 	DeploymentNotFoundCode    ErrorCode = "deploymentNotFoundErr"    // Could not find deployment to update

--- a/internal/events/errors_test.go
+++ b/internal/events/errors_test.go
@@ -5,7 +5,6 @@ package events
 import (
 	"errors"
 	"os"
-	"syscall"
 	"testing"
 	"time"
 
@@ -42,8 +41,6 @@ func (s *ErrorsSuite) TestGoErrorWithAttrs() {
 
 	s.NotEqual(time.Time{}, event.Time)
 	s.Equal("publish/restoreREnv/failure/unknown", event.Type)
-	s.Equal(syscall.Errno(2), event.Data["Err"])
-	s.Equal("/nonexistent", event.Data["Path"])
 }
 
 type testErrorDetails struct {
@@ -80,7 +77,5 @@ func (s *ErrorsSuite) TestErrorObjectAndDetails() {
 
 	s.NotEqual(time.Time{}, event.Time)
 	s.Equal("publish/restorePythonEnv/failure/serverErr", event.Type)
-	s.Equal(syscall.Errno(2), event.Data["Err"])
-	s.Equal("/nonexistent", event.Data["Path"])
 	s.Equal(500, event.Data["Status"])
 }

--- a/internal/publish/publish.go
+++ b/internal/publish/publish.go
@@ -127,6 +127,8 @@ func (p *defaultPublisher) emitErrorEvents(err error, log logging.Logger) {
 	dashboardURL := ""
 	directURL := ""
 
+	var data events.EventData
+
 	// Record the error in the deployment record
 	if p.Target != nil {
 		p.Target.Error = agentErr
@@ -139,14 +141,14 @@ func (p *defaultPublisher) emitErrorEvents(err error, log logging.Logger) {
 			dashboardURL = getDashboardURL(p.Account.URL, p.Target.ID)
 			directURL = getDirectURL(p.Account.URL, p.Target.ID)
 			agentErr.Data["dashboard_url"] = dashboardURL
+
+			mapstructure.Decode(publishFailureData{
+				DashboardURL: dashboardURL,
+				DirectURL:    directURL,
+			}, &data)
 		}
 	}
 
-	var data events.EventData
-	mapstructure.Decode(publishFailureData{
-		DashboardURL: dashboardURL,
-		DirectURL:    directURL,
-	}, &data)
 	maps.Copy(data, agentErr.GetData())
 
 	// Fail the phase

--- a/internal/types/error.go
+++ b/internal/types/error.go
@@ -44,7 +44,6 @@ func NewAgentError(code ErrorCode, err error, details any) *AgentError {
 	msg := ""
 	if err != nil {
 		msg = err.Error()
-		mapstructure.Decode(err, &data)
 	}
 	if details != nil {
 		detailMap := make(ErrorData)

